File: rust_exercises/leetcode/Problem726/Solution.java
class Solution {
    public String countOfAtoms(String formula) {
        Map<Integer, Integer> open2Time = new HashMap<>();
        Map<Integer, Integer> close2Skip = new HashMap<>();
        
        int i = 0;
        Stack<Integer> temp = new Stack<>();
        while (i < formula.length()) {
            char c = formula.charAt(i);
            
            i += 1;
            
            if (c == '(') {
                temp.push(i - 1);
            } else if (c == ')') {
                int time = 0;
                int j = i - 1;
                while (i < formula.length() && Character.isDigit(formula.charAt(i))) {
                    time = time * 10 + (int) (formula.charAt(i) - '0');
                    i += 1;
                }
                time = Math.max(1, time);
                open2Time.put(temp.pop(), time);
                close2Skip.put(j, i);
            }
        }
        
        Map<String, Integer> map = new TreeMap<>();
        int count = 0;
        int time = 1;
        StringBuilder sb = new StringBuilder();
        
        i = 0;
        while (i < formula.length()) {
            char c = formula.charAt(i);
            
            if (Character.isUpperCase(c) || c == '(' || c == ')') {
                if (sb.length() > 0) {
                    String element = sb.toString();
                    count = Math.max(count, 1);
                    map.put(element, map.getOrDefault(element, 0) + count * time);
                    
                    sb = new StringBuilder();
                    count = 0;
                }
            }
            
            if (c == ')') {
                time /= temp.pop();
                i = close2Skip.get(i);
            } else {
                if (c == '(') {
                    temp.push(open2Time.get(i));
                    time *= temp.peek();
                } else if (Character.isAlphabetic(c)) {
                    sb.append(c);
                } else if (Character.isDigit(c)) {
                    count = count * 10 + (int) (c - '0');
                }
                i += 1;
            }
        }
        
        if (sb.length() > 0) {
            String element = sb.toString();
            count = Math.max(count, 1);
            map.put(element, map.getOrDefault(element, 0) + count * time);
            sb = new StringBuilder();
        }
        
        for (Map.Entry<String, Integer> entry: map.entrySet()) {
            sb.append(entry.getKey());
            if (entry.getValue() > 1) {
                sb.append(Integer.toString(entry.getValue()));
            }
        }
        
        return sb.toString();
    }
}

File: rust_exercises/leetcode/Problem726/solution.rs
use std::collections::HashMap;
use std::collections::BTreeMap;
use std::cmp::max;

impl Solution {
    pub fn count_of_atoms(formula: String) -> String {
        let mut formula: Vec<char> = formula.as_bytes().iter().map(|x| *x as char).collect();
        
        let mut open2time: HashMap<u32, u32> = HashMap::new();
        let mut close2skip: HashMap<u32, u32> = HashMap::new();
        
        let mut i = 0;
        let mut temp: Vec<u32> = vec![];
        
        while i < formula.len() {
            let c = formula[i];
            
            i += 1;
            
            if c == '(' {
                temp.push(i as u32 - 1);
            } else if c == ')' {
                let mut time = 0;
                let j = i - 1;
                while i < formula.len() && formula[i].is_digit(10) {
                    time = time * 10 + formula[i].to_digit(10).unwrap();
                    i += 1;
                }
                time = max(1, time);
                open2time.insert(temp.pop().unwrap() as u32, time);
                close2skip.insert(j as u32, i as u32);
            }
        }
        
        let mut map = BTreeMap::new();
        let mut count = 0;
        let mut time = 1;
        let mut element = String::new();
        
        i = 0;
        
        while i < formula.len() {
            let c = formula[i];
            
            if c.is_uppercase() || c == '(' || c == ')' {
                if element.len() > 0 {
                    count = max(count, 1);
                    match map.get(&element) {
                        Some(x) => map.insert(element, x + count * time),
                        None => map.insert(element, count * time),
                    };
                    element = String::new();
                    count = 0;
                }
            }
            
            if c == ')' {
                time /= temp.pop().unwrap();
                i = *close2skip.get(&(i as u32)).unwrap() as usize;
            } else {
                if c == '(' {
                    temp.push(*open2time.get(&(i as u32)).unwrap());
                    time *= temp[temp.len() - 1];
                } else if c.is_alphabetic() {
                    element.push(c);
                } else if c.is_digit(10) {
                    count = count * 10 + c.to_digit(10).unwrap();
                }
                i += 1;
            }
        }
        
        if element.len() > 0 {
            count = max(count, 1);
            match map.get(&element) {
                Some(x) => map.insert(element, x + count * time),
                None => map.insert(element, count * time),
            };
        }
        
        let mut result = String::new();
        
        for (element, count) in map.iter() {
            result.push_str(&element);
            if *count > 1 {
                result.push_str(&count.to_string());
            }
        }
        
        result
    }
}
